File: adt2unit/pas/update_fine_effects.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure update_fine_effects(chan: Byte);

var
  eLo,eHi,
  eLo2,eHi2: Byte;

begin
  eLo  := LO(effect_table[chan]);
  eHi  := HI(effect_table[chan]);
  eLo2 := LO(effect_table2[chan]);
  eHi2 := HI(effect_table2[chan]);

  Case eLo of
    ef_ArpggVSlideFine:
      volume_slide(chan,eHi DIV 16,eHi MOD 16);

    ef_FSlideUpFine:
      portamento_up(chan,eHi,nFreq(12*8+1));

    ef_FSlideDownFine:
      portamento_down(chan,eHi,nFreq(0));

    ef_FSlUpVSlF:
      volume_slide(chan,eHi DIV 16,eHi MOD 16);

    ef_FSlDownVSlF:
      volume_slide(chan,eHi DIV 16,eHi MOD 16);

    ef_FSlUpFineVSlide:
      portamento_up(chan,fslide_table[chan],nFreq(12*8+1));

    ef_FSlUpFineVSlF:
      begin
        portamento_up(chan,fslide_table[chan],nFreq(12*8+1));
        volume_slide(chan,eHi DIV 16,eHi MOD 16);
      end;

    ef_FSlDownFineVSlide:
      portamento_down(chan,fslide_table[chan],nFreq(0));

    ef_FSlDownFineVSlF:
      begin
        portamento_down(chan,fslide_table[chan],nFreq(0));
        volume_slide(chan,eHi DIV 16,eHi MOD 16);
      end;

    ef_TPortamVSlideFine:
      volume_slide(chan,eHi DIV 16,eHi MOD 16);

    ef_Vibrato:
      If vibr_table[chan].fine then
        vibrato(chan);

    ef_Tremolo:
      If trem_table[chan].fine then
        tremolo(chan);

    ef_VibratoVolSlide:
      If vibr_table[chan].fine then
        vibrato(chan);

    ef_VibratoVSlideFine:
      begin
        volume_slide(chan,eHi DIV 16,eHi MOD 16);
        If vibr_table[chan].fine then
          vibrato(chan);
      end;

    ef_VolSlideFine:
      volume_slide(chan,eHi DIV 16,eHi MOD 16);

    ef_extended2+ef_fix2+ef_ex2_GlVolSlideUpF:
      global_volume_slide(eHi,BYTE_NULL);

    ef_extended2+ef_fix2+ef_ex2_GlVolSlideDnF:
      global_volume_slide(BYTE_NULL,eHi);
  end;

  Case eLo2 of
    ef_ArpggVSlideFine:
      volume_slide(chan,eHi2 DIV 16,eHi2 MOD 16);

    ef_FSlideUpFine:
      portamento_up(chan,eHi2,nFreq(12*8+1));

    ef_FSlideDownFine:
      portamento_down(chan,eHi2,nFreq(0));

    ef_FSlUpVSlF:
      volume_slide(chan,eHi2 DIV 16,eHi2 MOD 16);

    ef_FSlDownVSlF:
      volume_slide(chan,eHi2 DIV 16,eHi2 MOD 16);

    ef_FSlUpFineVSlide:
      portamento_up(chan,fslide_table2[chan],nFreq(12*8+1));

    ef_FSlUpFineVSlF:
      begin
        portamento_up(chan,fslide_table2[chan],nFreq(12*8+1));
        volume_slide(chan,eHi2 DIV 16,eHi2 MOD 16);
      end;

    ef_FSlDownFineVSlide:
      portamento_down(chan,fslide_table2[chan],nFreq(0));

    ef_FSlDownFineVSlF:
      begin
        portamento_down(chan,fslide_table2[chan],nFreq(0));
        volume_slide(chan,eHi2 DIV 16,eHi2 MOD 16);
      end;

    ef_TPortamVSlideFine:
      volume_slide(chan,eHi2 DIV 16,eHi2 MOD 16);

    ef_Vibrato:
      If vibr_table2[chan].fine then
        vibrato2(chan);

    ef_Tremolo:
      If trem_table2[chan].fine then
        tremolo2(chan);

    ef_VibratoVolSlide:
      If vibr_table2[chan].fine then
        vibrato2(chan);

    ef_VibratoVSlideFine:
      begin
        volume_slide(chan,eHi2 DIV 16,eHi2 MOD 16);
        If vibr_table2[chan].fine then
          vibrato2(chan);
      end;

    ef_VolSlideFine:
      volume_slide(chan,eHi2 DIV 16,eHi2 MOD 16);

    ef_extended2+ef_fix2+ef_ex2_GlVolSlideUpF:
      global_volume_slide(eHi2,BYTE_NULL);

    ef_extended2+ef_fix2+ef_ex2_GlVolSlideDnF:
      global_volume_slide(BYTE_NULL,eHi2);
  end;
end;

File: adt2unit/update_fine_effects.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// chan: 1..20
/*static*/ void update_fine_effects (uint8_t chan)
{
  chan--;

  uint8_t eLo  = effect_table[chan] & 0xFF;
  uint8_t eHi  = effect_table[chan] >> 8;
  uint8_t eLo2 = effect_table2[chan] & 0xFF;
  uint8_t eHi2 = effect_table2[chan] >> 8;

  switch (eLo)
  {
    case ef_ArpggVSlideFine:
      volume_slide (chan + 1, eHi / 16, eHi % 16);
      break;

    case ef_FSlideUpFine:
      portamento_up (chan + 1, eHi, nFreq (12*8+1));
      break;

    case ef_FSlideDownFine:
      portamento_down (chan + 1, eHi, nFreq (0));
      break;

    case ef_FSlUpVSlF:
      volume_slide (chan + 1, eHi / 16, eHi % 16);
      break;

    case ef_FSlDownVSlF:
      volume_slide (chan + 1, eHi / 16, eHi % 16);
      break;

    case ef_FSlUpFineVSlide:
      portamento_up (chan + 1, fslide_table[chan], nFreq (12*8+1));
      break;

    case ef_FSlUpFineVSlF:
      portamento_up (chan + 1, fslide_table[chan], nFreq (12*8+1));
      volume_slide (chan + 1, eHi / 16, eHi % 16);
      break;

    case ef_FSlDownFineVSlide:
      portamento_down (chan + 1, fslide_table[chan], nFreq (0));
      break;

    case ef_FSlDownFineVSlF:
      portamento_down (chan + 1, fslide_table[chan], nFreq (0));
      volume_slide (chan + 1, eHi / 16, eHi % 16);
      break;

    case ef_TPortamVSlideFine:
      volume_slide (chan + 1, eHi / 16, eHi % 16);
      break;

    case ef_Vibrato:
      if (vibr_table[chan].fine)
        vibrato (chan + 1);
      break;

    case ef_Tremolo:
      if (trem_table[chan].fine)
        tremolo (chan + 1);
      break;

    case ef_VibratoVolSlide:
      if (vibr_table[chan].fine)
        vibrato (chan + 1);
      break;

    case ef_VibratoVSlideFine:
      volume_slide (chan + 1, eHi / 16, eHi % 16);
      if (vibr_table[chan].fine)
        vibrato (chan + 1);
      break;

    case ef_VolSlideFine:
      volume_slide (chan + 1, eHi / 16, eHi % 16);
      break;

    case ef_Extended2 + ef_fix2 + ef_ex2_GlVolSlideUpF:
      global_volume_slide (eHi, BYTE_NULL);
      break;

    case ef_Extended2 + ef_fix2 + ef_ex2_GlVolSlideDnF:
      global_volume_slide (BYTE_NULL, eHi);
      break;

    default:
      break;
  }

  switch (eLo2)
  {
    case ef_ArpggVSlideFine:
      volume_slide (chan + 1, eHi2 / 16, eHi2 % 16);
      break;

    case ef_FSlideUpFine:
      portamento_up (chan + 1, eHi2, nFreq (12*8+1));
      break;

    case ef_FSlideDownFine:
      portamento_down (chan + 1, eHi2, nFreq (0));
      break;

    case ef_FSlUpVSlF:
      volume_slide (chan + 1, eHi2 / 16, eHi2 % 16);
      break;

    case ef_FSlDownVSlF:
      volume_slide (chan + 1, eHi2 / 16, eHi2 % 16);
      break;

    case ef_FSlUpFineVSlide:
      portamento_up (chan + 1, fslide_table2[chan], nFreq (12*8+1));
      break;

    case ef_FSlUpFineVSlF:
      portamento_up (chan + 1, fslide_table2[chan], nFreq (12*8+1));
      volume_slide (chan + 1, eHi2 / 16, eHi2 % 16);
      break;

    case ef_FSlDownFineVSlide:
      portamento_down (chan + 1, fslide_table2[chan], nFreq (0));
      break;

    case ef_FSlDownFineVSlF:
      portamento_down (chan + 1, fslide_table2[chan], nFreq (0));
      volume_slide (chan + 1, eHi2 / 16, eHi2 % 16);
      break;

    case ef_TPortamVSlideFine:
      volume_slide (chan + 1, eHi2 / 16, eHi2 % 16);
      break;

    case ef_Vibrato:
      if (vibr_table2[chan].fine)
        vibrato2 (chan + 1);
      break;

    case ef_Tremolo:
      if (trem_table2[chan].fine)
        tremolo2 (chan + 1);
      break;

    case ef_VibratoVolSlide:
      if (vibr_table2[chan].fine)
        vibrato2 (chan + 1);
      break;

    case ef_VibratoVSlideFine:
      volume_slide (chan + 1, eHi2 / 16, eHi2 % 16);
      if (vibr_table2[chan].fine)
        vibrato2 (chan + 1);
      break;

    case ef_VolSlideFine:
      volume_slide (chan + 1, eHi2 / 16, eHi2 % 16);
      break;

    case ef_Extended2 + ef_fix2 + ef_ex2_GlVolSlideUpF:
      global_volume_slide (eHi2, BYTE_NULL);
      break;

    case ef_Extended2 + ef_fix2 + ef_ex2_GlVolSlideDnF:
      global_volume_slide (BYTE_NULL, eHi2);
      break;

    default:
      break;
  }
}
